libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,881 @@
+#include "AP_SmartAudio.h"
+#include <AP_Math/crc.h>
+#include <GCS_MAVLink/GCS.h>
+
+#define SA_DEBUG     // DEV TESTS ACTIVATE
+#ifdef SA_DEBUG
+# define debug(fmt, args...)	hal.console->printf(" " fmt "\n", ##args)
+//#define debug(fmt, args...) gcs().send_text(MAV_SEVERITY_DEBUG , fmt, ##args);
+#else
+//# define debug(fmt, args...)	do {} while(0)
+#endif
+
+extern const AP_HAL::HAL &hal;
+const char *TAG="VTX-SMARTAUDIO ";
+
+
+// table of user settable parameters
+const AP_Param::GroupInfo AP_SmartAudio::var_info[] = {
+
+    // @Param: DEFAULTS
+    // @DisplayName: VTX will be configured with VTX retrieved from hardware vtx.
+    // @Description: VTX will be configured with VTX retrieved from hardware vtx.
+    // @Values: 0: Disabled, 1: Enabled
+    // @User: Advanced
+    AP_GROUPINFO("DEFAULTS", 3, AP_SmartAudio, _smart_audio_param_setup_defaults, 0),
+
+    AP_GROUPEND
+
+};
+
+AP_SmartAudio::AP_SmartAudio()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _singleton = this;
+
+}
+
+AP_SmartAudio *AP_SmartAudio::_singleton;
+
+// initialization start making a request settings to the vtx
+bool AP_SmartAudio::init()
+{
+
+    debug("SmartAudio init");
+
+    if (AP::vtx().get_enabled()==0){
+        debug("SmartAudio protocol it's not active");
+        return false;
+    }
+
+    // init uart
+    _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_SmartAudio, 0);
+    if (_port!=nullptr){
+
+          if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_SmartAudio::loop, void),
+                                          "SmartAudio",
+                                          512, AP_HAL::Scheduler::PRIORITY_IO, -1)) {
+            return false;
+            }
+
+        // setup port options
+        _port->set_stop_bits(2);
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+
+        // initialize AP_VideoTx settings
+        request_settings();
+
+        return true;
+    }
+    return false;
+}
+
+void AP_SmartAudio::loop(){
+
+    // initialise uart (this must be called from within tick b/c the UART begin must be called from the same thread as it is used from)
+    _port->begin(AP_SMARTAUDIO_UART_BAUD , AP_SMARTAUDIO_UART_BUFSIZE_RX, AP_SMARTAUDIO_UART_BUFSIZE_TX);
+
+
+     while (true) {
+
+        // now time to control loop switching
+        uint32_t now=AP_HAL::millis();
+
+        // when one packet is processed in the iteration will be true.
+        bool packet_processed=false;
+
+        // command to process
+        Packet current_command;
+
+        // setup a queue polling delay of 20 ms.
+        hal.scheduler->delay(50);
+
+
+
+        // Proccess response in the next 200 milis from the request are sent.
+        if (now-_last_request_sended_at_ms>100 && now-_last_request_sended_at_ms<=1000 && _is_waiting_response){
+            // allocate response buffer
+            uint8_t _response_buffer[AP_SMARTAUDIO_UART_BUFSIZE_RX];
```

Comment:
Have a look at how RCProtocol does this. I am not against allocating such a small buffer on the stack, but you should cope with not getting all the data at once. 
In RCProtocol we read bytes into a buffer until we have a header, if we don't have enough bytes we loop and sleep. You should do something similar instead of assuming that you have a complete packet in the UART.

Reply:
Fine, i'll check the actual approach which has in account the splitted responses, but seems that i have a bug on it. 
I'll research the sleep approach.

Reply:
After some rework is working with splitted responses, max uart buffer size is 16 bytes, max buffer retained 32 bytes.